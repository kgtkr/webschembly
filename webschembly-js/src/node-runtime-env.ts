import * as fs from "fs";
import * as path from "path";

export function createLogger({
  logDir = process.env.LOG_DIR || null,
  runtimeName = "untitled",
}) {
  const logBasename = path.basename(runtimeName) + "-" + Date.now();
  let logFile = null;
  if (logDir !== null) {
    try {
      fs.mkdirSync(logDir);
    } catch (e) {
      if (e.code !== "EEXIST") {
        throw e;
      }
    }

    logFile = fs.openSync(path.join(logDir, logBasename + ".log"), "a");
  }
  let instantiateCount = 0;

  return {
    instantiate: (buf) => {
      if (logDir !== null) {
        fs.writeFileSync(
          path.join(logDir, logBasename + "-" + instantiateCount + ".wasm"),
          buf
        );
      }
      instantiateCount++;
    },
    log: (s) => {
      if (logFile !== null) {
        fs.writeSync(logFile, s + "\n");
      }
    },
  };
}

export function createNodeRuntimeEnv({
  runtimeName = "untitled",
  exit = process.exit,
  logger = createLogger({ runtimeName }),
  runtimeModule = new WebAssembly.Module(
    fs.readFileSync(process.env["WEBSCHEMBLY_RUNTIME"])
  ),
  writeBuf = (fd, buf) => {
    switch (fd) {
      case 1:
        process.stdout.write(buf);
        break;
      case 2:
        process.stderr.write(buf);
        break;
      default:
        throw new Error(`Unsupported file descriptor: ${fd}`);
    }
  },
}) {
  return {
    runtimeName,
    exit,
    logger,
    runtimeModule,
    writeBuf,
  };
}
