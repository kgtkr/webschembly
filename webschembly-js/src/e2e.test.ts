import * as fs from "fs/promises";
import * as fsLegacy from "fs";
import { beforeAll, describe, expect, test } from "vitest";
import * as path from "path";
import {
  compilerConfigToString,
  createRuntime,
  type CompilerConfig,
} from "./runtime";
import { createNodeRuntimeEnv } from "./node-runtime-env";

function concatBufs(bufs: Uint8Array[]) {
  const bufLen = bufs.map((buf) => buf.length).reduce((a, b) => a + b, 0);
  const resultBuf = new Uint8Array(bufLen);
  let offset = 0;
  for (const buf of bufs) {
    resultBuf.set(buf, offset);
    offset += buf.length;
  }
  return resultBuf;
}

const snapshotDir = "e2e_snapshots";
const sourceDir = "fixtures";
const filenames = fsLegacy
  .readdirSync(sourceDir)
  .filter((file) => file.endsWith(".scm"));

const compilerConfigs: CompilerConfig[] = [
  {},
  { enableJitOptimization: false },
  { enableJit: false },
];

describe("E2E test", () => {
  let runtimeModule: WebAssembly.Module;
  beforeAll(async () => {
    runtimeModule = new WebAssembly.Module(
      await fs.readFile(process.env["WEBSCHEMBLY_RUNTIME"]!)
    );
  });

  describe.each(
    compilerConfigs.map((compilerConfig) => [
      compilerConfigToString(compilerConfig),
      compilerConfig,
    ])
  )("%s", (_, compilerConfig) => {
    describe.each(filenames)("%s", (filename) => {
      let srcBuf: Buffer;
      beforeAll(async () => {
        srcBuf = await fs.readFile(path.join(sourceDir, filename));
      });

      test(
        "snapshot test",
        async () => {
          let exitCode = 0;
          const stdoutBufs: Uint8Array[] = [];
          const stderrBufs: Uint8Array[] = [];
          const runtime = await createRuntime(
            await createNodeRuntimeEnv({
              runtimeName: filename,
              exit: (code) => {
                exitCode = code;
              },
              writeBuf: (fd, buf) => {
                switch (fd) {
                  case 1:
                    stdoutBufs.push(new Uint8Array(buf));
                    break;
                  case 2:
                    stderrBufs.push(new Uint8Array(buf));
                    break;
                  default:
                    throw new Error(`Unsupported file descriptor: ${fd}`);
                }
              },
              loadRuntimeModule: async () => runtimeModule,
            }),
            {
              compilerConfig,
            }
          );

          runtime.loadStdlib();
          runtime.loadSrc(srcBuf);
          runtime.cleanup();

          const stdout = new TextDecoder().decode(concatBufs(stdoutBufs));
          const stderr = new TextDecoder().decode(concatBufs(stderrBufs));

          await expect(exitCode).toMatchFileSnapshot(
            `${snapshotDir}/${filename}-exitCode`
          );
          await expect(stdout).toMatchFileSnapshot(
            `${snapshotDir}/${filename}-stdout`
          );
          await expect(stderr).toMatchFileSnapshot(
            `${snapshotDir}/${filename}-stderr`
          );
        },
        3 * 6 * 1000
      );
    });
  });
});
