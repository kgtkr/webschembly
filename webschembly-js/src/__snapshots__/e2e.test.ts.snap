// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`E2E test > add.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "3
",
}
`;

exports[`E2E test > builtin_some_name.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "30
3
",
}
`;

exports[`E2E test > closure.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "42
45
",
}
`;

exports[`E2E test > define.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": ""Hello, World!"
",
}
`;

exports[`E2E test > define_builtin.scm > snapshot test 1`] = `
{
  "exitCode": 1,
  "stderr": "set! builtin is not allowed
",
  "stdout": "",
}
`;

exports[`E2E test > define_in_let.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "1
2
3
",
}
`;

exports[`E2E test > display.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "(2 . (() . ("hello" . ((1 . (2 . ())) . (abc . (efg . (abc . ())))))))
",
}
`;

exports[`E2E test > get_builtin.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "3
",
}
`;

exports[`E2E test > global_duplication.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "2
2
3
3
4
4
",
}
`;

exports[`E2E test > if.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": ""True1"
"False2"
",
}
`;

exports[`E2E test > if_nest.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": ""1"
"8"
"10"
"16"
",
}
`;

exports[`E2E test > lambda.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "42
",
}
`;

exports[`E2E test > large_rec.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "0
",
}
`;

exports[`E2E test > let_in_global.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "1
2
3
4
1
2
",
}
`;

exports[`E2E test > let_in_local.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "1
2
3
4
1
2
",
}
`;

exports[`E2E test > list_syntax.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "()
(1 . (2 . (3 . (4 . ()))))
(1 . (2 . (3 . (4 . 5))))
(1 . (2 . (3 . (4 . ()))))
1
",
}
`;

exports[`E2E test > local_duplication.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "1
2
3
2
3
4
",
}
`;

exports[`E2E test > local_duplication2.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "2
3
",
}
`;

exports[`E2E test > mut_cell_args.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "1
2
2
",
}
`;

exports[`E2E test > not_definable.scm > snapshot test 1`] = `
{
  "exitCode": 1,
  "stderr": "CompilerError: Define is not allowed in this context: b
",
  "stdout": "",
}
`;

exports[`E2E test > num_cmp.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "#t
#f

#f
#f
#t

#f
#t
#f

#t
#f
#t

#t
#t
#f

",
}
`;

exports[`E2E test > rec.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "5050
",
}
`;

exports[`E2E test > set_builtin.scm > snapshot test 1`] = `
{
  "exitCode": 1,
  "stderr": "set! builtin is not allowed
",
  "stdout": "",
}
`;

exports[`E2E test > tak.b.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "7
",
}
`;

exports[`E2E test > vector.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "#()
#(1 2 3)
#(1 2 3)
2
#(1 4 3)
#(1 4 3)
",
}
`;
